### ISSUE TITLE
Basic Battleplan

### ISSUE THREAD

**Issue by JHeimbach** [BetterMood/better-mood#4]
Hey Guys,

I open this issue as a start to discuss a basic plan how we want to go forward.

Please add your ideas for this issue or open own issues with tasks.

**Comment by mistermuckle** [comment]
@nschmoyer has the good idea of moving some of the third party libs into composer, to help get the codebase down to a reasonable size. I’m for that if we do one library per commit, so we can revert if we get into trouble. Probably good to diff the composer version with the Moodle version before deleting, to make sure we aren’t deleting customizations.

**Comment by JHeimbach** [comment]
and we should split php and js libs

**Comment by nschmoyer** [comment]
@mistermuckle I agree. A lot of those libraries are extended, customized, or otherwise hacked up.

@JHeimbach yes, it is a damn disaster in that lib folder.

What do you guys think about coming up with a proposed project structure and adding it to the wiki? I feel like I'm flying blind right now.

**Comment by mistermuckle** [comment]
We could move the libs to custom repos, and load them via composer, just to get them out of the way.

For each lib,

1. Fork the unmodified version of the lib
2. Apply changes mad by Moodle
3. Add the fork to composer.json

**Comment by CarwynNelson** [comment]
Personally I wouldn't want to think about modifying the entire project structure until we were more comfortable with what is already there.

The existing project structure already exists to achieve a goal (modularity?) and so I think we should be cognisant of that going forward.

I have posted this on discord - but after getting some of the basic cleaning tasks out of the way (composer, local dev etc) I think we should look to tackle one small, simple module within Moodle and refactor that to our best ability. I think that will give some key insights to be able to move forward.

I also think we should look to get the current unit tests running on a ci build. I believe one is already configured in this codebase so we should just use that.

I know the tests aren't that great - but they are something and they should help us in not making too many breaking changes to the codebase.

With that said - I don't think we should use any of the existing phpunit infrastructure if we can avoid it. It does test the functionality of Moodle and I can respect that (it solved a problem at a time), but it requires you to bring the kitchen sink just to write a simple test. If the code was written to be testable then all this "bending over backwards" to write tests wouldn't be required.

I'm going to raise an issue to get local unit tests running on the docker build. Hopefully in an easy one-command kind of way. We'll see how that goes.

Longer term I think it would be nice to split the front-end and the back-end into separate apps (even if they are in the same repo). There already seems to be a decent selection of webservices contained within Moodle so I don't think it is a giant leap to build a new front-end separate to the codebase the replaces the default one. I think in order to keep that one simple we limit it to just the user-facing features and keep the admin dashboards as they are.

Again that is longer term and I think we should keep that as a something to aim for (if everyone agrees with) rather than something we are working towards right now.

Bit of a long comment but hopefully there are some helpful bits in there.
